### ISSUE TITLE
DLQ Integration Test: Logstash should not increment a metric with a negative value

### ISSUE THREAD

**Issue by robbavey** [elastic/logstash#8168]
During the running of the dlq integration test, occasionally Travis fails with the following error:

```
Exception in thread "Ruby-0-Thread-10@[main]>worker0: :1" java.lang.IllegalArgumentException: Counters can not be incremented by negative values
	at org.logstash.instrument.metrics.counter.LongCounter.<clinit>(org/logstash/instrument/metrics/counter/LongCounter.java:15)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(java/lang/Class.java:348)
	at org.jruby.javasupport.JavaSupportImpl.loadJavaClass(org/jruby/javasupport/JavaSupportImpl.java:154)
	at org.jruby.javasupport.Java.getProxyClassOrNull(org/jruby/javasupport/Java.java:940)
	at org.jruby.javasupport.Java.getProxyClassOrNull(org/jruby/javasupport/Java.java:927)
	at org.jruby.javasupport.Java.getProxyOrPackageUnderPackage(org/jruby/javasupport/Java.java:885)
	at org.jruby.javasupport.JavaPackage.method_missing(org/jruby/javasupport/JavaPackage.java:259)
	at org.jruby.javasupport.JavaPackage$INVOKER$i$method_missing.call(org/jruby/javasupport/JavaPackage$INVOKER$i$method_missing.gen)
	at RUBY.<main>(:1)
	at org.jruby.Ruby.runInterpreter(org/jruby/Ruby.java:839)
	at org.jruby.Ruby.loadFile(org/jruby/Ruby.java:2876)
	at org.jruby.RubyKernel.requireCommon(org/jruby/RubyKernel.java:968)
	at org.jruby.RubyKernel.require(org/jruby/RubyKernel.java:961)
	at org.jruby.RubyKernel$INVOKER$s$1$0$require19.call(org/jruby/RubyKernel$INVOKER$s$1$0$require19.gen)
	at RUBY.<main>(:1)
	at org.jruby.Ruby.runInterpreter(org/jruby/Ruby.java:839)
	at org.jruby.Ruby.loadFile(org/jruby/Ruby.java:2876)
	at org.jruby.RubyKernel.requireCommon(org/jruby/RubyKernel.java:968)
	at org.jruby.RubyKernel.require(org/jruby/RubyKernel.java:961)
	at org.jruby.RubyKernel$INVOKER$s$1$0$require19.call(org/jruby/RubyKernel$INVOKER$s$1$0$require19.gen)
	at org.jruby.Ruby.runInterpreter(org/jruby/Ruby.java:839)
	at org.jruby.Ruby.loadFile(org/jruby/Ruby.java:2876)
	at org.jruby.RubyKernel.requireCommon(org/jruby/RubyKernel.java:968)
	at org.jruby.RubyKernel.require(org/jruby/RubyKernel.java:961)
	at org.jruby.RubyKernel$INVOKER$s$1$0$require19.call(org/jruby/RubyKernel$INVOKER$s$1$0$require19.gen)
	at org.jruby.Ruby.runInterpreter(org/jruby/Ruby.java:839)
	at org.jruby.Ruby.loadFile(org/jruby/Ruby.java:2876)
	at org.jruby.RubyKernel.requireCommon(org/jruby/RubyKernel.java:968)
	at org.jruby.RubyKernel.require(org/jruby/RubyKernel.java:961)
	at org.jruby.RubyKernel$INVOKER$s$1$0$require19.call(org/jruby/RubyKernel$INVOKER$s$1$0$require19.gen)
	at org.jruby.Ruby.runInterpreter(org/jruby/Ruby.java:839)
	at org.jruby.Ruby.loadFile(org/jruby/Ruby.java:2876)
	at org.jruby.RubyKernel.requireCommon(org/jruby/RubyKernel.java:968)
	at org.jruby.RubyKernel.require(org/jruby/RubyKernel.java:961)
	at org.jruby.RubyKernel$INVOKER$s$1$0$require19.call(org/jruby/RubyKernel$INVOKER$s$1$0$require19.gen)
	at org.jruby.Ruby.runInterpreter(org/jruby/Ruby.java:839)
	at org.jruby.Ruby.loadFile(org/jruby/Ruby.java:2876)
	at org.jruby.RubyKernel.requireCommon(org/jruby/RubyKernel.java:968)
	at org.jruby.RubyKernel.require(org/jruby/RubyKernel.java:961)
	at org.jruby.RubyKernel$INVOKER$s$1$0$require19.call(org/jruby/RubyKernel$INVOKER$s$1$0$require19.gen)
	at home.travis.build.elastic.logstash.build.logstash_minus_7_dot_0_dot_0_minus_alpha1_minus_SNAPSHOT.lib.bootstrap.environment.<main>(/home/travis/build/elastic/logstash/build/logstash-7.0.0-alpha1-SNAPSHOT/lib/bootstrap/environment.rb:70)
	at java.lang.invoke.MethodHandle.invokeWithArguments(java/lang/invoke/MethodHandle.java:627)
	at org.jruby.Ruby.runScript(org/jruby/Ruby.java:827)
	at org.jruby.Ruby.runNormally(org/jruby/Ruby.java:746)
	at org.jruby.Ruby.runNormally(org/jruby/Ruby.java:764)
	at org.jruby.Ruby.runFromMain(org/jruby/Ruby.java:577)
	at org.jruby.Main.doRunFromMain(org/jruby/Main.java:417)
	at org.jruby.Main.internalRun(org/jruby/Main.java:305)
	at org.jruby.Main.run(org/jruby/Main.java:232)
	at org.jruby.Main.main(org/jruby/Main.java:204)
```

One of the metrics is being incremented with a negative value, although the log message and stack trace gives no indication as to which one.

**Comment by jakelandis** [comment]
@robbavey - Can we close this as fixed ?

**Comment by robbavey** [comment]
@jakelandis Yeah, I think so - the change to nanos should do that. Closing

**Comment by redbaron4** [comment]
Hi @robbavey . Was the proposed fix part of 6.0.0 release (I am guessing not)?

I have been hit with this issue in production (using 6.0.0). Apparently this happens if the time on node jumps back (which is possible if NTP is running). I have [a thread](https://discuss.elastic.co/t/counters-can-not-be-incremented-by-negative-values-error-when-using-metric-filters-freezes-workers-and-corrupts-the-persistent-queue/108644) about this on elastic discussion board documenting stack traces and the impact.

**Comment by robbavey** [comment]
@redbaron4 This will be included in the next releases of Logstash, 6.0.1 and 6.1.0

**Comment by redbaron4** [comment]
@robbavey Thanks for the clarification.

Looking at the commit which addresses this, I see that Logstash will now use `nanoTime` for computing the elapsed time. In my case I am reasonably sure (from the `messages` logs posted on discussion thread) the error happens when NTP causes time to jump "back", thereby causing `@metric_events_time.increment` to be passed a negative value (`current_time` may be less than `start_time`). Wouldn't that problem still possibly occur even if we use `nanoTime`?
